File: 136. Single Number/136. Single Number.py
#
# Problem: 136. Single Number
# Difficulty: Easy
# Link: https://leetcode.com/problems/single-number/submissions/1996274291/
# Language: python
# Date: 2026-05-06


class Solution(object):
    def singleNumber(self, nums):
        xor_val = 0
        for i in nums:
            xor_val = xor_val ^ i

        return xor_val
        

File: 136. Single Number/136. Single Number.cpp
/*
 * Problem: 136. Single Number
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/single-number/
 * Language: cpp
 * Date: 2026-05-06
 */

// OPTIMAL APPROACH USING XOR

class Solution {
public:
    int singleNumber(vector<int>& nums) {

        int xor_val = 0;

        for (int i = 0; i < nums.size(); i++) {

            // Same numbers cancel each other in XOR
            // a ^ a = 0
            // 0 ^ x = x
            xor_val = xor_val ^ nums[i];
        }

        // Only unique element remains after all cancellations
        return xor_val;
    }
};
